### PR TITLE
Add patch ordering to assessment and installation patch summaries

### PR DESCRIPTION
**Pull request by benank**
Adds patch ordering to assessment and installation patch summaries.

Patches are ordered by the following criteria:
 - First by classification: Security, Critical, Other, Unclassified
 - Then by patchInstallationState (for install operations): Failed, Installed, Available, Pending, Excluded, NotSelected

### CODE REVIEW

File: src/core/src/service_interfaces/StatusHandler.py [hunk 1]
```diff
@@ -124,8 +124,51 @@ def set_package_assessment_status(self, package_names, package_versions, classif
...
+        }
+
+        patch_state_order = {
+            Constants.FAILED: 1,
+            Constants.INSTALLED: 2,
+            Constants.AVAILABLE: 3,
+            Constants.PENDING: 4,
+            Constants.EXCLUDED: 5,
+            Constants.NOT_SELECTED: 6
+        }
```
Comment (rane-rajasi) [inline src/core/src/service_interfaces/StatusHandler.py:150]: These 2 can be constants instead of a function

Comment (benank) [reply src/core/src/service_interfaces/StatusHandler.py]: Done

File: src/core/src/bootstrap/Constants.py [hunk 2]
```diff
@@ -275,3 +275,20 @@ class EnvLayer(EnumBackport):
         PRIVILEGED_OP_REBOOT = PRIVILEGED_OP_MARKER + "Reboot_Exception"
         PRIVILEGED_OP_EXIT = PRIVILEGED_OP_MARKER + "Exit_"
 
+    # Package / Patch State Ordering Constants
```
Comment (kjohn-msft) [inline src/core/src/bootstrap/Constants.py:278]: Expand comment to an additional line to describe why this is being done.

Comment (kjohn-msft) [reply src/core/src/bootstrap/Constants.py:278]: Maybe even encapsulate that in the names. As an example: PackageClassificationOrderInStatusReporting   (if the variable is not directly used too many times).

File: src/core/src/service_interfaces/StatusHandler.py [hunk 3]
```diff
@@ -112,7 +112,7 @@ def set_package_assessment_status(self, package_names, package_versions, classif
                 if patch_id == self.__assessment_packages[i]['patchId']:
                     patch_already_saved = True
                     self.__assessment_packages[i]['classifications'] = [classification]
-                    self.__assessment_packages[i]['patchState'] = status
+                    # self.__assessment_packages[i]['patchState'] = status
```
Comment (kjohn-msft) [inline src/core/src/service_interfaces/StatusHandler.py:115]: Why?

File: src/core/src/service_interfaces/StatusHandler.py [hunk 4]
```diff
@@ -124,8 +124,35 @@ def set_package_assessment_status(self, package_names, package_versions, classif
...
+            lowest_classification = Constants.PackageClassificationOrder[x["classifications"][0]]
+            for i in range(1, len(x["classifications"])):
+                lowest_classification = min(lowest_classification, Constants.PackageClassificationOrder[x["classifications"][i]])
+            return lowest_classification
+
+        # Sort by patch state first then sort by classification so each type of classification is already sorted at the end
+        list_sorted_by_patch_state = sorted(packages_list, key=sort_patch_state_key)
+        list_sorted_by_classification = sorted(list_sorted_by_patch_state, key=sort_classification_key)
+        # COMMENT: these sort calls can be combined into one by separating the value ranges in the tables above, such as classification_order to 100, 200, 300, etc
+        # and having one key function that combines the result of classification_order and patch_state_order
```
Comment (kjohn-msft) [inline src/core/src/service_interfaces/StatusHandler.py:153]: Check note: Is the outcome protected by a UT given dependency on specific behavior of sorted

File: src/core/src/service_interfaces/StatusHandler.py (hunk 3 above)
Comment (benank) [reply src/core/src/service_interfaces/StatusHandler.py:115]: `patchState` is deprecated based on the OS-agnostic design spec. On line 123, you can see that it is also commented out, and this matches it. No code in the repo currently uses `patchState`.

File: src/core/src/bootstrap/Constants.py (hunk 2 above)
Comment (benank) [reply src/core/src/bootstrap/Constants.py:278]: Will do both

File: src/core/src/service_interfaces/StatusHandler.py (hunk 4 above)
Comment (benank) [reply src/core/src/service_interfaces/StatusHandler.py:153]: Yes

File: src/core/tests/Test_CoreMain.py [hunk 5]
```diff
@@ -439,16 +439,16 @@ def test_install_all_packages_for_centos_autopatching(self):
...
+        self.assertEqual(json.loads(substatus_file_data[1]["formattedMessage"]["message"])["patches"][2]["name"], "selinux-policy-targeted.noarch")
+        self.assertTrue("Other" in str(json.loads(substatus_file_data[1]["formattedMessage"]["message"])["patches"][2]["classifications"]))
         self.assertTrue("Installed" == json.loads(substatus_file_data[1]["formattedMessage"]["message"])["patches"][2]["patchInstallationState"])
+        self.assertEqual(json.loads(substatus_file_data[1]["formattedMessage"]["message"])["patches"][0]["name"], "libgcc.i686")
+        self.assertTrue("libgcc.i686_4.8.5-28.el7_CentOS Linux_7.9.2009" in str(json.loads(substatus_file_data[1]["formattedMessage"]["message"])["patches"][0]["patchId"]))
+        self.assertTrue("Security" in str(json.loads(substatus_file_data[1]["formattedMessage"]["message"])["patches"][0]["classifications"]))
+        self.assertTrue("Installed" == json.loads(substatus_file_data[1]["formattedMessage"]["message"])["patches"][0]["patchInstallationState"])
         self.assertTrue(substatus_file_data[2]["name"] == Constants.PATCH_METADATA_FOR_HEALTHSTORE)
```
Comment (kjohn-msft) [inline src/core/tests/Test_CoreMain.py:452]: Please summarize the significance of the changes in the tests (only for easing the review; not a request for any comment in code).

File: src/extension/tests/helpers/AssessmentSummary.json [hunk 6]
```diff
@@ -0,0 +1,72 @@
+{
+    "assessmentActivityId": "d7e26b52-c18e-46ec-a44b-8cb7bd7a62a3",
```
Comment (kjohn-msft) [inline src/extension/tests/helpers/AssessmentSummary.json:2]: Please qualify the file name with the specific test it's intended to support. We don't want future test code lazy-leaning on test artifacts meant for specific tests resulting in cascading test failures if something needs to be tweaked for one test (or a closely related group of tests). Same comment for the other json.

File: src/core/tests/Test_StatusHandler.py [hunk 7]
```diff
@@ -358,6 +358,34 @@ def test_set_current_operation_auto_assess_non_assessment(self):
...
+            self.assertEqual(installation_patches_sorted[2]["name"], "test-package-11")
+            self.assertEqual(installation_patches_sorted[3]["name"], "test-package-10")
+            self.assertEqual(installation_patches_sorted[4]["name"], "test-package-9")
+            self.assertEqual(installation_patches_sorted[5]["name"], "test-package-8")
+            self.assertEqual(installation_patches_sorted[6]["name"], "test-package-7")
+            self.assertEqual(installation_patches_sorted[7]["name"], "test-package-5")
+            self.assertEqual(installation_patches_sorted[8]["name"], "test-package-4")
+            self.assertEqual(installation_patches_sorted[9]["name"], "test-package-3")
+            self.assertEqual(installation_patches_sorted[10]["name"], "test-package-2")
+            self.assertEqual(installation_patches_sorted[11]["name"], "test-package-1")
```
Comment (kjohn-msft) [inline src/core/tests/Test_StatusHandler.py:387]: It's a little bit of work now but for long-term ease of reading, please append the classification and state as a comment at the end of each line to make clear what the outcome was. Ideally tabbed correctly so it reads like a table, and helps for fast comparison against the input json.

File: src/core/src/service_interfaces/StatusHandler.py (hunk 4 above)
Comment (kjohn-msft) [reply src/core/src/service_interfaces/StatusHandler.py:153]: Verified in principle - closing comment.

File: src/core/tests/Test_CoreMain.py (hunk 5 above)
Comment (benank) [reply src/core/tests/Test_CoreMain.py:452]: The tests in this file were changed based on the new patch assessment/installation summary ordering:
 - First by classification: Security, Critical, Other, Unclassified
 - Then by patchInstallationState (for install operations): Failed, Installed, Available, Pending, Excluded, NotSelected

The changes here modify the indices to match the new expected ordering of the patches in the report summary.

File: src/core/tests/Test_StatusHandler.py (hunk 7 above)
Comment (benank) [reply src/core/tests/Test_StatusHandler.py]: Good idea, added!

**Review by kjohn-msft — APPROVED** [review]

**Pull request merged**